src/App.tsx

```diff
@@ -1,26 +1,353 @@
+        todo.id === todoId ? { ...todo, isSubmitting: true } : todo,
+      ),
+    );
+
+    // Perform the deletion on the server
+    return deleteTodo(todoId)
+      .then(() => {
+        // On success, remove the todo from the state
+        setTodos(currentTodos =>
+          currentTodos.filter(todo => todo.id !== todoId),
+        );
+      })
+      .catch(() => {
+        // On error, revert the todo state and show the error message
+        setTodos(currentTodos =>
+          currentTodos.map(todo =>
+            todo.id === todoId ? { ...todo, isSubmitting: false } : todo,
+          ),
+        );
+        setErrorMessage('Unable to delete a todo');
+        setTimeout(() => {
+          setErrorMessage('');
+        }, 3000);
+        //throw error; // Propagate error for debugging
+      })
+      .finally(() => {
+        setIsSubmitting(false);
+        inputRef.current?.focus();
+      });
+  }
+
+  const handleTitleChange = (todoId: number, newTitle: string) => {
+    // Find the todo item by its ID
+    const todo = todos.find(t => t.id === todoId);
+
+    if (!todo) {
+      return; // If the todo doesn't exist, do nothing
+    }
+
+    if (!newTitle.trim()) {
+      deleteThisTodo(todoId);
+
+      return;
+    }
+
+    // Only proceed if the title has actually changed
+    if (todo.title === newTitle) {
+      return; // No need to update if the title is the same
+    }
+
+    // Mark the todo as submitting to show a loader for the specific todo
+    setTodos(prevTodos =>
+      prevTodos.map(t => (t.id === todoId ? { ...t, isSubmitting: true } : t)),
+    );
+
+    // Perform the title change on the server
+    return updateTodo({ ...todo, title: newTitle })
+      .then(() => {
+        // Update the local state with the new title after success
+        setTodos(prevTodos =>
+          prevTodos.map(t =>
+            t.id === todoId
+              ? { ...t, title: newTitle, isSubmitting: false }
+              : t,
+          ),
+        );
+
+        // Reset the editing state and selected todo id
+        setIsEdited(false);
+        setSelectedTodoId(null);
+      })
+      .catch(() => {
+        // Handle error by resetting the submitting state and showing an error message
+        setErrorMessage('Unable to update a todo');
+
+        setTodos(prevTodos =>
+          prevTodos.map(t =>
+            t.id === todoId ? { ...t, isSubmitting: false } : t,
+          ),
+        );
+
+        setTimeout(() => {
+          setErrorMessage(''); // Reset error message after 3 seconds
+        }, 3000);
+      });
+  };
+
+  const submitChangedTitle = (e, todo) => {
+    e.preventDefault();
+    const newTitle = e.target[0].value.trim(); // Get the value from the input field
+
+    if (newTitle !== todo.title) {
+      // Only update if the title has changed
+      handleTitleChange(todo.id, newTitle); // Call handleTitleChange with the new title
+    }
+  };
+
+  function addTodo(event: React.FormEvent) {
```

Comment:
It is important to stick to one way of creating functions
```suggestion
  const addTodo = (event: React.FormEvent) => {
```